### PR TITLE
Fixes for `tensornet` backend may not pick up MPI

### PR DESCRIPTION
**Pull request by 1tnguyen**
### Description

The `tensornet` backend propagates MPI to `cutensornet` lib during ctor. However, the static instance might be created before MPI initialization; hence need to be reset to pick up MPI and propagate to `cutensornet`.

### CODE REVIEW

File: runtime/nvqir/cutensornet/simulator_tensornet_register.cpp [hunk 1]
```diff
@@ -20,20 +23,45 @@ class SimulatorTensorNet : public SimulatorTensorNetBase {
...
-    if (cudaq::mpi::is_initialized())
+    if (cudaq::mpi::is_initialized()) {
       resetCuTensornetComm(m_cutnHandle);
+      m_cutnMpiInitialized = false;
+    }
   }
+
+private:
+  friend nvqir::CircuitSimulator * ::getCircuitSimulator_tensornet();
+  /// @brief Have we initialized cuTensorNet MPI during ctor?
```
Comment (schweitzpgi) [inline runtime/nvqir/cutensornet/simulator_tensornet_register.cpp:45]:
```suggestion
  /// @brief Has cuTensorNet MPI been initialized?
```

File: runtime/nvqir/cutensornet/simulator_tensornet_register.cpp [hunk 2]
```diff
@@ -20,20 +23,45 @@ class SimulatorTensorNet : public SimulatorTensorNetBase {
...
+  bool m_cutnMpiInitialized = false;
 };
 } // namespace nvqir
 
 /// Register this Simulator class with NVQIR under name "tensornet"
-NVQIR_REGISTER_SIMULATOR(nvqir::SimulatorTensorNet, tensornet)
+extern "C" {
```
Comment (schweitzpgi) [inline runtime/nvqir/cutensornet/simulator_tensornet_register.cpp:51]: Q: Why is the extern "C" needed? Python entry point?

File: runtime/nvqir/cutensornet/simulator_tensornet_register.cpp [hunk 3]
```diff
@@ -20,20 +23,45 @@ class SimulatorTensorNet : public SimulatorTensorNetBase {
     // Note: this requires CUTENSORNET_COMM_LIB as described in
     // the Getting Started section of the cuTensorNet library documentation
     // (Installation and Compilation).
-    if (cudaq::mpi::is_initialized())
+    if (cudaq::mpi::is_initialized()) {
       initCuTensornetComm(m_cutnHandle);
+      m_cutnMpiInitialized = true;
```
[the remark this answers is not in the document]
Comment (1tnguyen) [reply runtime/nvqir/cutensornet/simulator_tensornet_register.cpp:28]: [cutensornet](https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/api/functions.html#distributed-parallelization-api) doesn't have public APIs to query the state of the distributed interface (e.g., whether a communicator has been assigned). It only exposes a few APIs to set the communicator and query MPI-related information (rank/size) which would require the communicator.

File: runtime/nvqir/cutensornet/simulator_tensornet_register.cpp (hunk 2 above)
Comment (1tnguyen) [reply runtime/nvqir/cutensornet/simulator_tensornet_register.cpp:51]: These are plugin-style `dlsym` hooks (`runtime/common/PluginUtils.h`) for simulators/platforms hence need to be `extern "C"`.
The runtime looks for these symbols to load the simulator/platform.

**Pull request merged**